File: solutions/fortran/sieve/1/sieve.f90
module sieve
  implicit none

contains

 function primes(limit) result(array)
    integer, intent(in) :: limit
    integer, allocatable :: array(:)
    integer :: k 
    integer, allocatable:: numbers(:)
    integer :: sqrt_limit
    integer :: n 
    if (limit > 1) then
        allocate(numbers, source=[(k,k=1,limit)])
        n = 2
        sqrt_limit = int(sqrt(real(limit)))
      do while (n<= sqrt_limit)
        do k=n*n,limit,n
          numbers(k) = 0
        end do
        n = n+1
        do while (n==0) 
          n= n+1
        end do
      end do
      array = pack(numbers, numbers>1)
    else
      allocate(array(0))
    end if
  end function primes


end module sieve

File: solutions/javascript/sieve/1/sieve.js
//
// classical sieve

export const primes = (N) => {
  if(N<2) return [];
  let numbers = Array.from(Array(N-1), (_,k)=>k+2);
  let n=2
  while(n<=N){
    for(let k=n*n; k<=N; k+=n){
      numbers[k-2]=0;
    }  
    n++;
  }
  return numbers.filter(p=>p!=0);
};
